Introduce support for running phased actions

lsmaira opened the pull request:
This commit introduces the ability of running multiple build actions in
different phases of the build. These actions are passed by the client
through the tooling api.

With this commit, a single action can be added to each one of the
supporting phases (after projects are loaded, after projects are
evaluated and after tasks are run).

This feature allows improvements like running actions that call a model
builder modifying the graph tasks, and then it is possible to first
fetch a model and then execute tasks, in this order. e.g. Android Studio
sync + source generation.

Signed-off-by: Lucas Smaira <lsmaira@google.com>

### Context
This change will be useful for Android Studio to sync and generate sources in a single Gradle invocation.

Problem and solution discussed with Gradlers in private meetings instead of public forums. Design doc included.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes -> updated java docs
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation
- [x] Recognize contributor in release notes

lsmaira commented:
I made a few changes concerning the reviewed doc:
- Instead of having an afterSettingsEvaluatedAction, I replaced that with afterProjectsLoaded action, since we kind of want projects information (to ask user to select what projects to sync in the future)
- Added a forTasks() method to PhasedActionExecuter to allow predefined tasks to be configured by the client

lsmaira commented on subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiClasspathIntegrationTest.groovy:
```diff
@@ -30,7 +30,7 @@ class ToolingApiClasspathIntegrationTest extends AbstractIntegrationSpec {
         resolve.classpath.size() == 2
         resolve.classpath.any {it.name ==~ /slf4j-api-.*\.jar/}
-        resolve.classpath.find { it.name ==~ /gradle-tooling-api.*\.jar/ }.size() < 1.7 * 1024 * 1024
+        resolve.classpath.find { it.name ==~ /gradle-tooling-api.*\.jar/ }.size() < 1.8 * 1024 * 1024
```
I'm not sure if this is the right "fix" for this test...
My understanding is that this test checks the size of the jar and it is normal to have it increased from time to time since code is added to it. Is that right?

oehme replied:
Yes that's right - This is just to make sure it is a minified JAR and we don't accidentally package unnecessary stuff in there. The failure serves as a reminder to ask "Did I really add that much code?"

lsmaira commented:
Hey Stefan,
I didn't really understand these failing tests. They seem unrelated. Is there a fair issue in my commit or was it an infrastructure problem?

oehme commented:
The [compatibility check](https://builds.gradle.org/repository/download/Gradle_Check_SanityCheck/11324397:id/report-distributions-binary-compatibility-report.html#) failed because new APIs weren't marked with `@since` and `@Incubating`. Can you have a look?

lsmaira commented:
Sure, done! Sorry, I hadn't found the logs of the tests. Now I see where they are.

oehme commented:
Next failure is just a bad import I think: https://builds.gradle.org/viewLog.html?buildId=11486801&buildTypeId=Gradle_Check_Stage_QuickFeedback_Trigger

lsmaira commented:
Is there anything I should do before the code review? Or does the flag "waiting-for:contributor" means anything else?

blindpirate commented:
@lsmaira Sorry I didn't see the recent commit you pushed and thought @oehme was waiting for your feedback. Changed the tag now.

oehme commented on subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientProvidedPhasedActionRunner.java:
```diff
@@ -0,0 +1,132 @@
+
+        addBuildListener(phasedAction, buildController);
+
+        // We don't know if the model builders invoked in the after configuration action will add tasks to the graph or not
```
This should be improved in `buildController` by providing an API similar to

```
buildController.build {
  loadSettings()
  configureBuild()
  runTasks()
}
```

That method would automatically fire the right buildStarted/buildFinished hooks and you would get a failure if you try to do things in the wrong order. Using such an API you could run the build only up to the point that you want and react to what happened in between.

This can be a follow-up improvement, but until this is done the `projectsLoaded` hook is kinda unhelpful, because it would still configure everything even though the user only wanted information based on the settings file.

Will you have time to look into this after we merge this change?

lsmaira replied:
Thank you for the suggestion :) I was thinking about a solution for that.
I'll try to implement it in another commit in this same pull request soon, so it doesn't block review for now and hopefully we can merge together.

lsmaira replied:
@oehme, I looked into this and implemented a solution based on this suggestion. As it is kind of orthogonal to this change, I thought it was better to submit as a separate PR (it touches progress events to make Loading projects and Configure Projects two separate build operations).

The commit is [here](https://github.com/lsmaira/gradle/tree/build-controller-api) and once we merge this PR, I'll rebase on top of master and send another PR.

Is it okay?

oehme commented on subprojects/tooling-api-builders/src/test/groovy/org/gradle/tooling/internal/provider/runner/ClientProvidedPhasedActionRunnerTest.groovy:
```diff
@@ -0,0 +1,199 @@
+import org.gradle.tooling.internal.provider.serialization.SerializedPayload
+import spock.lang.Specification
+
+class ClientProvidedPhasedActionRunnerTest extends Specification {
```
❌ This comment applies to all unit tests: Can you please reduce the number of Mocks to only those interactions you are trying to verify? Everything else should be Stubs (or just the real implementation if it is a value object). 

The reason behind this is that mocks lock you down to a very specific implementation, so that refactorings can become extremely painful.

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java:
```diff
@@ -0,0 +1,155 @@
+         *
+         * @since 4.7
+         */
+        <T> Builder addAfterLoadingAction(BuildAction<T> buildAction, ResultHandler<? super T> handler) throws MultipleBuildActionsException;
```
⭕️ `projectsLoaded()`

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java:
```diff
@@ -0,0 +1,155 @@
+         *
+         * @since 4.7
+         */
+        <T> Builder addAfterConfigurationAction(BuildAction<T> buildAction, ResultHandler<? super T> handler) throws MultipleBuildActionsException;
```
⭕️ `projectsEvaluated()`

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java:
```diff
@@ -0,0 +1,155 @@
+         *
+         * @since 4.7
+         */
+        <T> Builder addAfterBuildAction(BuildAction<T> buildAction, ResultHandler<? super T> handler) throws MultipleBuildActionsException;
```
⭕️ `buildFinished()`

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java:
```diff
@@ -0,0 +1,155 @@
+        /**
+         * Adds the given action and its result handler to be executed after projects are loaded.
+         *
+         * <p>The action or model builders invoked by it are run when plugins have not yet be applied, and so it should try to get only gradle default models.
```
⭕️ You can add model builders in an init script or with a settings plugin, so this comment is not accurate. It should say something along the lines of "models contributed by project plugins won't be available"

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java:
```diff
@@ -0,0 +1,155 @@
+    interface Builder {
+
+        /**
+         * Adds the given action and its result handler to be executed after projects are loaded.
```
⭕️ I would say "Executes the given action ....". Since this is not a list, so it's not really "adding".

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java:
```diff
@@ -0,0 +1,155 @@
+        /**
+         * Adds the given action and its result handler to be executed after tasks are run.
+         *
+         * <p>The action should not invoke task graph modifiers, since tasks have already been run.
```
⭕️ "task graph modifier" is not a term we use. I think I would leave this out. It's also kind of obvious from the fact that this is called "buildFinished". I don't think anyone would expect a modification of the StartParameter to have any effect at that point.

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java:
```diff
@@ -0,0 +1,155 @@
+        /**
+         * Adds the given action and its result handler to be executed after projects are configured and before tasks are run.
+         *
+         * <p>The action has already access to plugins' model builders. These model builders can modify the task graph, since tasks have not yet been run.
```
❌ Let's not encourage people to do this. It's the only way to solve your use case right now, but I think we want a stronger model than mutating the StartParameter in the future.

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/exceptions/MultipleBuildActionsException.java:
```diff
@@ -0,0 +1,33 @@
+ * @since 4.7
+ */
+@Incubating
+public class MultipleBuildActionsException extends RuntimeException {
```
❌ This is overly specific in my opinion. This is a programming error, not something that should be specially reacted to. A simple IllegalArgumentException should do.

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/AfterLoadingResult.java:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @since 4.7
+ */
+public interface AfterLoadingResult<T> extends PhasedActionResult<T> {
```
⭕️ Why 3 different types instead of having a `getPhase()`?

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/PhasedActionResult.java:
```diff
@@ -0,0 +1,51 @@
+     * @since 4.7
+     */
+    @Nullable
+    Throwable getFailure();
```
❌ Individual actions should not have failures, that just complicates the user code unnecessarily. If an action encounters a recoverable exception it should handle it itself. If an action throws an exception, Gradle should fail the build immediately.

As a result, only once failure callback is necessary and the individual result listeners don't need a failure callback.

oehme commented:
@lsmaira I'm on vacation this week, but will give this another review early next week so we can get it merged. The target release will be Gradle 4.8.

lsmaira commented:
Alright, I'll send a commit changing mentions of 4.7 to 4.8.

lsmaira commented:
Just updated target version to 4.8 and rebased everything to avoid later conflicts (like moving to StartParameterInternal) and take advantage that master is already in 4.8

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/PhasedActionResult.java:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return The result. {@code null} if a failure occurred.
+     *
+     * @since 4.8
```
These methods don't need a since tag, the information is already on the class.

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/PhasedActionResult.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Gets the action result if it completed successfully.
+     *
+     * @return The result. {@code null} if a failure occurred.
```
❌ This shouldn't happen. If the action failed the build should fail and no result should be returned to the phased handler

oehme commented:
I'm afraid just checking whether some tasks were defined is not enough. Think about a phased action which just wants to look at the projects after the settings were evaluated, but doesn't want to configure them/run tasks at all. How would Gradle distinguish that from an action that does want Gradle to run the default tasks even though it only has a `projectsLoaded` hook? 

It seems like the action should explicitly define up to which phase Gradle should run. That could be `execute tasks` by default, but could be changed to `just configure projects` or `just load the settings`. I'm okay if we do this in a follow-up, but until then we need to document that the current API will always execute the default tasks if none are defined.

Regarding your comments about the individual failures: I think I confused two different things:

1. the actions run inside the daemon
2. the listeners receiving the results on the client side

If 1. fails, the build should terminate immediately. So if I provide an action that hooks into the projectsLoaded hook and that action fails, then Gradle should not configure or run tasks. We should have a test for that.

Failures in 2. should indeed be rethrown only after the build itself has finished.

lsmaira commented:
Got it! That makes sense and I had completely missed interrupting the build if an action fails (I had only guaranteed that future actions wouldn't be executed).

I removed the last commit concerning listener exception handling and sent a new commit addressing your comments:

- I documented in PhasedActionExecuter that default tasks will be run in case none is specified. I also added a test to guarantee that default tasks are not run (it is, of course, ignored for now, and I'll enable it in my next PR, where I'll add, as you suggested, a upToPhase API when running a PhasedAction)

- Exceptions in client provided listeners are rethrown after the build has finished

- Now the build will immediately fail if one of the actions fails (e.g. if projectsLoadedAction fails projects will not be configured). I added a test in PhasedBuildActionCrossVersionSpec for that. To give more meaning to the exceptions, I unwrap them in ProviderConnection to be sent back to the ToolingAPI so they match the expected type. I think that a better fix, however, would be to check for failures in the API that will be introduced in BuildController, so we don't wrap and after unwrap the exceptions, instead we just don't execute future phases. I plan to do that in the next PR, if you're okay with that.

oehme commented on subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ProviderConnection.java:
```diff
@@ -137,22 +143,48 @@ public Object runClientAction(Object clientAction, BuildCancellationToken cancel
+        try {
+            return run(action, cancellationToken, listenerConfig, new PhasedActionEventConsumer(failsafePhasedActionResultListener, payloadSerializer, listenerConfig.buildEventConsumer),
+                providerParameters, params);
+        } catch (BuildExceptionVersion1 e) {
```
⭕️ I think this logic should live in the phased build action runner (i.e. a try-catch around buildController.run())

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/PhasedBuildActionExecuter.java:
```diff
@@ -0,0 +1,142 @@
+    /**
+     * Specifies the tasks to execute before executing the BuildFinishedAction and after the ProjectsEvaluatedAction.
+     *
+     * The graph task can be changed in model builders invoked. If not configured, null, or an empty array is passed, then no tasks will be executed unless one of the model builders configures it.
```
❌ This is not true. Tasks will always be run. I think we should remove this paragraph and instead explain task execution later when we add the "upToPhase" concept. Also let's not call out the possibility of mutating the start parameters in the model builder.

oehme commented:
Thanks @lsmaira, this looks very good. I'll take it from here.

oehme commented:
@lsmaira One more request: Could you rebase your branch on master?

lsmaira commented:
Sorry for the delay, @oehme, I had some meetings the entire day. I rebased on top of master and made sure that :toolingApi:check, :toolingApiBuilders:check, :launcher:check and :core:check passed.

lsmaira commented:
@oehme is this performance test failure related to the PR? Looking at the test history it looks like it passed by an instability period. I couldn't run the test locally

oehme commented:
Yeah I think you got a bad commit form master when you rebaselined, can you try again?

oehme commented:
Thank you again @lsmaira! Looking forward to the `upTo(Stage)` PR :)
